control-plane/p4RuntimeSerializer.cpp

```diff
@@ -1069,9 +1069,16 @@ class P4RuntimeEntriesConverter {
             protoParam->set_param_id(parameterId++);
             auto parameter = actionDecl->parameters->parameters.at(parameterIndex);
             auto width = parameter->type->width_bits();
-            auto value = stringRepr(arg->expression->to<IR::Constant>()->value, ROUNDUP(width, 8));
-            if (value == boost::none) continue;
-            protoParam->set_value(*value);
+            if (arg->expression->is<IR::Constant>()) {
+                auto value = stringRepr(arg->expression->to<IR::Constant>(), ROUNDUP(width, 8));
```

Comment:
And I don't understand the question.
But it made me realize that `stringReprConstant` wasn't correct because I should check that the bitwidth of the provided literal doesn't exceed the bitwidth of the type. And that I only really need the number of bytes when calling `mpz_export`.

Reply:
:)
The question is why round up to bytes, and not keep the defined width of the constant value, which may be any number of bits including non-multiples of 8. Are we not supporting literals in binary form so that we can express any width?

Reply:
The P4Info message includes the exact bitwidth. Here we are talking about some literals that are provided as values for these entities (parameter / match field). They are represented in P4Runtime as binary strings and are byte-aligned, with the extra bits (in most-significant position) being zeroed-out. When the P4Runtime server executes the INSERT operation for these static entries it verifies that the provided binary string matches the P4Info bitwidth (i.e. that there are no extra bits).